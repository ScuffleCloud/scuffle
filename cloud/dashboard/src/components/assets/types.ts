export type TimelineDotVariant = "default" | "live" | "finished";
