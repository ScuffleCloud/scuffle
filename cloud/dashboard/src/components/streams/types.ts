// Define types
export type VideoStatus = "live" | "finished";

export const DEFAULT_LOGIN_MODE: LoginMode = "login";

export type LoginMode =
    | "login" // Default login mode - magic-link
    | "password"
    | "passkey"
    | "magic-link-sent"
    | "forgot-password"
    | "password-reset-sent";

export interface VideoStream {
    id: string;
    name: string;
    status: VideoStatus;
    created: string;
    relatedStreams?: VideoStream[];
}

export interface StreamEvent {
    id: string;
    name: string;
    status: VideoStatus;
    created: string;
}

export type ChartData = {
    eventData: {
        name: string;
        type: string;
        value: number[];
    }[];
    lineData: {
        timestamp: number;
        value: number;
    }[];
};
