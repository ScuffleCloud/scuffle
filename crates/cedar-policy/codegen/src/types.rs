use std::collections::{BTreeMap, BTreeSet};
use std::fmt::Write;

use cedar_policy_core::ast::Id;
use cedar_policy_core::validator::RawName;

use crate::utils::{to_snake_ident, to_upper_camel_ident};

/// Represents a reference to a Cedar type with its namespace
#[derive(PartialEq, PartialOrd, Eq, Ord, Hash, Debug)]
pub(crate) struct CedarRef {
    pub(crate) id: Id,
    pub(crate) namespace: NamespaceId,
}

impl From<RawName> for CedarRef {
    fn from(value: RawName) -> Self {
        let qualified_value = value.qualify_with(None);
        Self {
            id: qualified_value.basename().clone(),
            namespace: NamespaceId {
                items: qualified_value.namespace_components().cloned().collect(),
            },
        }
    }
}

impl CedarRef {
    /// Converts this reference to a CedarType
    pub(crate) fn into_cedar_ty(self) -> CedarType {
        match self.id.as_ref() {
            "Bool" => CedarType::Bool,
            "Long" => CedarType::Long,
            "String" => CedarType::String,
            _ => CedarType::Reference(self),
        }
    }

    /// Gets the identifier path for this reference
    pub(crate) fn ident_path(&self) -> Vec<syn::Ident> {
        self.namespace
            .items
            .iter()
            .map(to_snake_ident)
            .chain(std::iter::once(&self.id).map(to_upper_camel_ident))
            .collect()
    }
}

/// Represents a namespace identifier with components
#[derive(PartialEq, PartialOrd, Eq, Ord, Hash, Default, Clone)]
pub(crate) struct NamespaceId {
    pub items: Vec<Id>,
}

/// Represents a field within a Cedar record type
#[derive(Debug)]
pub(crate) struct CedarTypeStructField {
    pub ty: CedarType,
    pub optional: bool,
}

/// Represents an action entity identifier
#[derive(Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub(crate) struct ActionEid {
    pub id: Option<CedarRef>,
    pub name: String,
}

/// Represents different Cedar types
#[derive(Debug)]
pub(crate) enum CedarType {
    String,
    Long,
    Bool,
    Record {
        fields: BTreeMap<String, CedarTypeStructField>,
        allows_additional: bool,
    },
    Set(Box<CedarType>),
    Enum(BTreeSet<String>),
    Reference(CedarRef),
    Entity {
        parents: Vec<CedarRef>,
        shape: Box<CedarType>,
        tag_type: Option<Box<CedarType>>,
    },
}

impl CedarType {
    /// Returns true if this type represents an entity
    pub(crate) fn is_entity(&self) -> bool {
        matches!(self, Self::Entity { .. } | Self::Enum(_))
    }
}

impl std::fmt::Display for ActionEid {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(id) = &self.id {
            id.fmt(f)?;
            f.write_str("::")?;
        }

        f.write_char('"')?;
        f.write_str(&self.name)?;
        f.write_char('"')
    }
}

impl std::fmt::Debug for NamespaceId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "NamespaceId(\"{self}\")")
    }
}

impl std::fmt::Display for CedarRef {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.namespace.fmt(f)?;
        if !self.namespace.items.is_empty() {
            f.write_str("::")?;
        }
        self.id.fmt(f)
    }
}

impl std::fmt::Display for NamespaceId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut first = true;
        for part in &self.items {
            if !first {
                f.write_str("::")?;
            }
            part.fmt(f)?;
            first = false;
        }
        Ok(())
    }
}
