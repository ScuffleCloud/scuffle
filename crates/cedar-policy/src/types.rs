use serde::ser::SerializeMap;

/// A type used by [crate::CedarEntity::TagType] to indicate that no tag is provided.
// The reason this is an enum is because an enum with no variants is unconstructable.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NoTag {}

impl serde::Serialize for NoTag {
    fn serialize<S>(&self, _: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        // Since this is an empty enum it can never be constructed.
        match *self {}
    }
}

/// A type used by [crate::CedarAction::Context] to indicate that no context is provided.
// This is a struct because we need to be able to construct this.
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct EmptyContext;

impl serde::Serialize for EmptyContext {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_map(Some(0))?.end()
    }
}

/// A type used by [crate::CedarEntity::Attrs] to indicate that the entity has no attributes.
// This is a struct because we need to be able to construct this.
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct NoAttributes;

impl serde::Serialize for NoAttributes {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_map(Some(0))?.end()
    }
}
