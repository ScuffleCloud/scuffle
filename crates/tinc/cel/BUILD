load("//build/utils:package.bzl", "scuffle_package", "scuffle_test")
load("//build/utils:cargo_toml.bzl", "cargo_toml")

cargo_toml()

scuffle_package(
    crate_name = "tinc-cel",
    compile_data = [
        ":Cargo.toml",
    ],
    test = scuffle_test(
        insta = True,
    ),
)
