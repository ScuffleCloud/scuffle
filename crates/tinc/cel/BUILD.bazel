load("//misc/utils/rust:package.bzl", "scuffle_package", "scuffle_test")
load("//misc/utils/rust:manifest.bzl", "cargo_toml")

cargo_toml()

scuffle_package(
    crate_name = "tinc-cel",
    compile_data = [
        ":Cargo.toml",
    ],
    test = scuffle_test(
        insta = True,
    ),
)
