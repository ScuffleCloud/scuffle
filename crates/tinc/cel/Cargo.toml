[package]
name = "tinc-cel"
version = "0.1.2"
edition = "2024"
description = "A CEL Type used by tinc-build and tinc for validating protobuf messages"
repository = "https://github.com/scufflecloud/scuffle"
authors = ["Scuffle <opensource@scuffle.cloud>"]
readme = "README.md"
documentation = "https://docs.rs/tinc-build"
license = "MIT OR Apache-2.0"
keywords = ["grpc", "protobuf", "tonic", "codegen"]

[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = ['cfg(coverage_nightly)'] }

[features]
## Enables runtime enum type indexing
runtime = []
## Enables documentation of feature flags
docs = ["dep:document-features"]

[dependencies]
bytes = "1.5"
chrono = "0.4"
document-features = { optional = true, version = "0.2" }
email_address = "0.2.9"
float-cmp = "0.10"
fmtools = "0.1"
linkme = "0.3"
num-traits = "0.2.19"
regex = "1"
scuffle-workspace-hack.workspace = true
thiserror = "2"
url = "2"
uuid = "1"

[dev-dependencies]
insta = "1.43"

[package.metadata.xtask.powerset]
additive-features = ["docs"]

[package.metadata.docs.rs]
all-features = true
rustdoc-args = [
  "--cfg",
  "docsrs",
  "--sort-modules-by-appearance",
  "--generate-link-to-definition",
]

[package.metadata.cargo-sync-rdme.badge]
style = "flat-square"

[package.metadata.cargo-sync-rdme.badge.badges]
license = true
docs-rs = true
crates-io = true
github-actions = { workflows = "ci.yaml" }
codecov = true

[package.metadata.xtask.release]
group = "tinc"
