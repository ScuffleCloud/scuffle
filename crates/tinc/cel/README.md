<!-- dprint-ignore-file -->
<!-- sync-readme title [[ -->
# tinc-cel
<!-- sync-readme ]] -->

> [!WARNING]  
> This crate is under active development and may not be stable.

<!-- sync-readme badge [[ -->
[![docs.rs](https://img.shields.io/docsrs/tinc-cel/0.2.0.svg?logo=docs.rs&label=docs.rs&style=flat-square)](https://docs.rs/tinc-cel/0.2.0)
[![crates.io](https://img.shields.io/badge/crates.io-v0.2.0-orange?style=flat-square&logo=rust&logoColor=white)](https://crates.io/crates/tinc-cel/0.2.0)
![License: MIT OR Apache-2.0](https://img.shields.io/badge/license-MIT%20OR%20Apache--2.0-purple.svg?style=flat-square)
![Crates.io Size](https://img.shields.io/crates/size/tinc-cel/0.2.0.svg?style=flat-square)
![Crates.io Downloads](https://img.shields.io/crates/dv/tinc-cel/0.2.0.svg?&label=downloads&style=flat-square)
[![Codecov](https://img.shields.io/codecov/c/github/scufflecloud/scuffle.svg?label=codecov&logo=codecov&style=flat-square)](https://app.codecov.io/gh/scufflecloud/scuffle)
<!-- sync-readme ]] -->

---

<!-- sync-readme rustdoc [[ -->
Currently this is a fully private api used by `tinc` and `tinc-build` to
compile and execute [CEL](https://cel.dev/) expressions.

### Feature flags

* **`runtime`** —  Enables runtime enum type indexing
* **`docs`** —  Enables documentation of feature flags

### License

This project is licensed under the MIT or Apache-2.0 license.
You can choose between one of them if you use this work.

`SPDX-License-Identifier: MIT OR Apache-2.0`
<!-- sync-readme ]] -->
