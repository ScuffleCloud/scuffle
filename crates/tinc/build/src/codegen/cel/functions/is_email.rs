use syn::parse_quote;
use tinc_cel::CelValue;

use super::Function;
use crate::codegen::cel::compiler::{CompileError, CompiledExpr, CompilerCtx, ConstantCompiledExpr};
use crate::codegen::cel::types::CelType;
use crate::types::{ProtoType, ProtoValueType};

#[derive(Debug, Clone, Default)]
pub(crate) struct IsEmail;

// this.isEmail(arg) -> arg in this
impl Function for IsEmail {
    fn name(&self) -> &'static str {
        "isEmail"
    }

    fn syntax(&self) -> &'static str {
        "<this>.isEmail()"
    }

    fn compile(&self, ctx: CompilerCtx) -> Result<CompiledExpr, CompileError> {
        let Some(this) = &ctx.this else {
            return Err(CompileError::syntax("missing this", self));
        };

        if !ctx.args.is_empty() {
            return Err(CompileError::syntax("does not take any arguments", self));
        }

        let this = this.clone().into_cel()?;

        match this {
            CompiledExpr::Constant(ConstantCompiledExpr { value }) => {
                Ok(CompiledExpr::constant(CelValue::cel_is_email(value)?))
            }
            this => Ok(CompiledExpr::runtime(
                CelType::Proto(ProtoType::Value(ProtoValueType::Bool)),
                parse_quote! {{
                    ::tinc::__private::cel::CelValue::cel_is_email(
                        #this,
                    )?
                }},
            )),
        }
    }
}

#[cfg(test)]
#[cfg(feature = "prost")]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use syn::parse_quote;
    use tinc_cel::CelValue;

    use crate::codegen::cel::compiler::{CompiledExpr, Compiler, CompilerCtx};
    use crate::codegen::cel::functions::{Function, IsEmail};
    use crate::codegen::cel::types::CelType;
    use crate::extern_paths::ExternPaths;
    use crate::path_set::PathSet;
    use crate::types::{ProtoType, ProtoTypeRegistry, ProtoValueType};

    #[test]
    fn test_is_email_syntax() {
        let registry = ProtoTypeRegistry::new(crate::Mode::Prost, ExternPaths::new(crate::Mode::Prost), PathSet::default());
        let compiler = Compiler::new(&registry);
        insta::assert_debug_snapshot!(IsEmail.compile(CompilerCtx::new(compiler.child(), None, &[])), @r#"
        Err(
            InvalidSyntax {
                message: "missing this",
                syntax: "<this>.isEmail()",
            },
        )
        "#);

        insta::assert_debug_snapshot!(IsEmail.compile(CompilerCtx::new(compiler.child(), Some(CompiledExpr::constant(CelValue::String("not-an-email".into()))), &[])), @r"
        Ok(
            Constant(
                ConstantCompiledExpr {
                    value: Bool(
                        false,
                    ),
                },
            ),
        )
        ");

        insta::assert_debug_snapshot!(IsEmail.compile(CompilerCtx::new(compiler.child(), Some(CompiledExpr::constant(CelValue::String("troy@scuffle.cloud".into()))), &[])), @r"
        Ok(
            Constant(
                ConstantCompiledExpr {
                    value: Bool(
                        true,
                    ),
                },
            ),
        )
        ");

        insta::assert_debug_snapshot!(IsEmail.compile(CompilerCtx::new(compiler.child(), Some(CompiledExpr::constant(CelValue::List(Default::default()))), &[
            cel_parser::parse("1 + 1").unwrap(), // not an ident
        ])), @r#"
        Err(
            InvalidSyntax {
                message: "does not take any arguments",
                syntax: "<this>.isEmail()",
            },
        )
        "#);
    }

    #[test]
    #[cfg(not(valgrind))]
    fn test_is_email_runtime() {
        let registry = ProtoTypeRegistry::new(crate::Mode::Prost, ExternPaths::new(crate::Mode::Prost), PathSet::default());
        let compiler = Compiler::new(&registry);

        let string_value =
            CompiledExpr::runtime(CelType::Proto(ProtoType::Value(ProtoValueType::String)), parse_quote!(input));

        let output = IsEmail
            .compile(CompilerCtx::new(compiler.child(), Some(string_value), &[]))
            .unwrap();

        insta::assert_snapshot!(postcompile::compile_str!(
            postcompile::config! {
                test: true,
                dependencies: vec![
                    postcompile::Dependency::version("tinc", "*"),
                ],
            },
            quote::quote! {
                fn is_email(input: &str) -> Result<bool, ::tinc::__private::cel::CelError<'_>> {
                    Ok(#output)
                }

                #[test]
                fn test_is_email() {
                    assert_eq!(is_email("troy@scuffle.cloud").unwrap(), true);
                    assert_eq!(is_email("not-an-email").unwrap(), false);
                }
            },
        ));
    }
}
