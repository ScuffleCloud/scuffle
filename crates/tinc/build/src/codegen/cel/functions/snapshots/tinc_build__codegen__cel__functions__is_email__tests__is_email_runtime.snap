---
source: crates/tinc/build/src/codegen/cel/functions/is_email.rs
expression: "postcompile::compile_str!(postcompile::config!\n{\n    test: true, dependencies:\n    vec![postcompile::Dependency::version(\"tinc\", \"*\"),],\n}, quote::quote!\n{\n    fn is_email(input: &str) -> Result<bool,\n    ::tinc::__private::cel::CelError<'_>> { Ok(#output) } #[test] fn\n    test_is_email()\n    {\n        assert_eq!(is_email(\"troy@scuffle.cloud\").unwrap(), true);\n        assert_eq!(is_email(\"not-an-email\").unwrap(), false);\n    }\n},)"
---
exit status: 0
--- test_stdout
running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in [ELAPSED]s
--- expanded
#![feature(prelude_import)]
#[macro_use]
extern crate std;
#[prelude_import]
use std::prelude::rust_2024::*;
fn is_email(input: &str) -> Result<bool, ::tinc::__private::cel::CelError<'_>> {
    Ok({
        ::tinc::__private::cel::CelValue::cel_is_email(
            ::tinc::__private::cel::CelValueConv::conv(input),
        )?
    })
}
